lib/thread.c

```diff
@@ -1669,6 +1726,37 @@ void thread_call(struct thread *thread)
 	unsigned long helper;
 #endif
 	RUSAGE_T before, after;
+	struct timeval wkptime, diff;
+	uint64_t wkpintvl_usecs;
+
+	/*
+	 * The following code measures interval between consecutive
+	 * wakeup of the various threads under the current process
+	 * and log that in order to detect any possible CPU starvations.
+	 * Offcourse it can not be used to differentiate between
+	 * wether a process was indeed ready but it could not be
+	 * scheduled (due to other process hogging CPU) and that of
+	 * the process really being idle for long.
+	 *
+	 * In cases if a deamon is supposed to run some timely regular
+	 * jobs, the data captured here can definitely highlight if
+	 * the process ever went through CPU starvation or not.
+	 */
+	frr_get_realtime(&wkptime);
+	frr_with_mutex(&thread->master->mtx) {
```

Comment:
Grabing a mutex on *every* thread_call is really really expensive.  If you really need this make it a configure option so that people not using this will not be impacted, or find a way to safely update data via some lock free mech.

Reply:
I did anticipate this comment :) And if the concerns are really that pressing I am fine to add it an API for turning it on per-daemon basis. By default it could be off. And specific daemons can turn it on at compile-time. Does that sound good? Or do you expect it to be turned on dynamically at runtime.

In my opinion this should be at least configurable compile-time otherwise we won't have a way to detect CPU starvation/hog after any daemon runs into any such condition.

Reply:
the concerns are that pressing.  There is no need for that complicated of a solution.  one compile time option to turn it on is sufficient.

Reply:
Ok will put an API that can be used by interested daemons to turn it on during initialisation.

Reply:
Done. Added a thread_set_stats_collection() API to enable/disable extra thread statistics collection on a per-process basis. Though this has been kept as disabled by default in this PR.